File: P_4/Examen_P4/Napakalaki_Ruby_P4/lib/monster.rb
#encoding: utf-8
# MADE BY: JOSE MIGUEL HDEZ GARCIA & JAIRO ABRIL MOYA


module NapakalakiRuby
    
require_relative 'bad_consequence'
require_relative 'prize'
require_relative 'treasure_kind'
require_relative 'numeric_bad_consequence'
require_relative 'specific_bad_consequence'
require_relative 'death_bad_consequence'

class Monster
    
    attr_reader :name
    attr_reader :combatLevel
    attr_reader :bc
    attr_reader :prize
    attr_reader :level_change_against_cultist_player
    
    private_class_method :new

    # Constructor
    def initialize(name = "", combatLevel = 0, bc = BadConsequence.new(), prize = Prize.new(), lc)
        @name = name
        @combatLevel = combatLevel
        @bc = bc
        @prize = prize
        @level_change_against_cultist_player = lc
    end
    
    def self.new_normal_monster(name = "", combatLevel = 0, bc = BadConsequence.new(), 
                                prize = Prize.new(), lc)
        new(name, combatLevel, bc, prize, lc)
    end
    
    def self.new_cultist_monster(name = " ", combatLevel = 0, bc = BadConsequence.new(),
                                prize = Prize.new(), lc)
        new(name, combatLevel, bc, prize, lc)
    end
    
    def get_combat_level_against_cultist_player
        return (@level_change_against_cultist_player + @combatLevel)
    end
    
    def get_levels_gained
        @prize.level
    end
     
    def get_treasures_gained
        @prize.treasures
    end
    
    # Metodo to_string que pinta en pantalla lo deseado
    def to_s
        "\nNombre: #{@name}
        Nivel: #{@combatLevel.to_s}
        Mal rollo: #{@bc.to_s}
        #{@prize.to_s}"
    end
end
end

File: P_4/Napakalaki_Java_P4/src/napakalaki/Monster.java
// MADE BY: JOSE MIGUEL HDEZ GARCIA & JAIRO ABRIL MOYA

package napakalaki;


public class Monster
{
    // Declaración de los atributos PRIVADOS a utilizar
    private final String name;
    private final int combatLevel;
    private final Prize prize;
    private final BadConsequence badconsequence;

    private int levelChangeAgainstCultistPlayer;

    // Constructor que inicializa a valores por defecto
    public Monster(String name, int level, BadConsequence bc, Prize prize, int lc)
    {
        this.name = name;
        this.combatLevel = level;
        this.prize = prize;
        this.badconsequence = bc;
        this.levelChangeAgainstCultistPlayer = lc;
    }

    // Consultores básicos
    // Devuelve el nombre del monstruo
    public String getName(){
        return name;
    }

    // Devuelve el nivel de combate del monstruo
    public int getCombatLevel(){
        return combatLevel;
    }

    // Devuelve el buen rollo del monstruo
    public Prize getPrize(){
        return prize;
    }

    // Devuelve el mal rollo de monstruo
    public BadConsequence getBadConsequence(){
        return badconsequence;
    }

    // Devuelve los niveles que se han ganado al derrotar al monstruo
    public int getLevelsGained(){
        return prize.getLevel();
    }

    // Devuelve los tesoros que se han ganado al derrotar al monstruo
    public int getTreasuresGained(){
        return prize.getTreasures();
    }

    // Devuelve el nivel de combate contra los sectarios
    public int getCombatLevelAgainstCultistPlayer(){
        return this.levelChangeAgainstCultistPlayer + this.combatLevel;
    }

    // Método toString para pintar una secuencia
    @Override
    public String toString(){
        return "\n\nMonster{" + "\nName = " + name
            + " \nCombat Level = " + combatLevel
            + " \nPrize = " + prize
            + " \nBad Consequence = " + badconsequence + '}';
    }
}
